File: pdoo/ruby/spacestation.rb
#encoding:utf-8

require_relative "./damage.rb"
require_relative "./hangar.rb"
require "./lib/SpaceStationToUI.rb"
require_relative "./loot.rb"
require_relative "./shieldbooster.rb"
require_relative "./shotresult.rb"
require_relative "./suppliespackage.rb"
require_relative "./weapon.rb"
require_relative "./lib/CardDealer.rb"
require_relative 'transformation'

module Deepspace
  # Estación espacial
  class SpaceStation
    # Cantidad máxima de combustible de una estación espacial
    @@MAXFUEL = 100
    # Potencia de escudo perdido por cada unidad de potencia de disparo
    @@SHIELDLOSSPERUNITSHOT = 0.1
    
    attr_reader :name, :ammoPower, :fuelUnits, :shieldPower, :nMedals, :weapons, :shieldBoosters, :pendingDamage, :hangar

    def initialize(n,supplies)
      @name = n
      @ammoPower = supplies.ammoPower
      @fuelUnits = [@@MAXFUEL,supplies.fuelUnits].min
      @shieldPower = supplies.shieldPower
      @nMedals = 0
      @weapons = Array.new
      @shieldBoosters = Array.new
      @pendingDamage = nil
      @hangar = nil
    end

    def newCopy(station)
      @name = station.name
      @ammoPower = station.ammoPower
      @fuelUnits = station.fuelUnits
      @shieldPower = station.shieldPower
      @nMedals = station.nMedals
      @weapons = station.weapons
      @shieldBoosters = station.shieldBoosters
      @pendingDamage = station.pendingDamage
      @hangar = station.hangar
    end

    # Asigna un nuevo valor de combustible
    private def assignFuelValue(f)
      if f<@@MAXFUEL
        @fuelUnits = f
      end
    end
    
    # Elimina el daño si no tiene efecto
    private def cleanPendingDamage()
      if @pendingDamage.hasNoEffect
        @pendingDamage = nil
      end
    end

    # Elimina las armas y potenciadores de escudos con 0 usos
    def cleanUpMountedItems
      @weapons.each_with_index do |weapon,i|
        if weapon.uses==0
          @weapons.delete_at(i)
        end
      end
      @shieldBoosters.each_with_index do |shield,i|
        if shield.uses==0
          @shieldBoosters.delete_at(i)
        end
      end
    end
  
    # Elimina el hangar
    def discardHangar
      @hangar = nil
    end

    # Elimina un escudo de la nave
    def discardShieldBooster(i)
      if i>=0 and i<@shieldBoosters.length
        if @pendingDamage!=nil
          @pendingDamage.discardShieldBooster
        end
        @shieldBoosters.delete_at(i)
      end
    end
    
    # Elimina un arma de la nave
    def discardWeapon(i)
      if i>=0 and i<@weapons.length
        if @pendingDamage!=nil
          w = @weapons[i]
          @pendingDamage.discardWeapon(w)
          cleanPendingDamage
        end
        @weapons.delete_at(i)
      end
    end

    # Elimina el potenciador de escudo con índice 'i' del hangar
    def discardShieldBoosterInHangar(i)
      if @hangar!=nil
        @hangar.removeShieldBooster(i)
      end
    end

    # Elimina el arma con índice 'i' del hangar
    def discardWeaponInHangar(i)
      if @hangar!=nil
        @hangar.removeWeapon(i)
      end
    end

    # Calcula la potencia del disparo
    def fire
      factor = 1.0
      @weapons.each do |weapon|
        factor = factor*weapon.useIt
      end
      @ammoPower*factor
    end

    # Devuelve la velocidad de la estación como fracción de combustible
    def getSpeed
      @fuelUnits/@@MAXFUEL
    end

    # Equipa el escudo de índice 'i' del hangar a la estación
    def mountShieldBooster(i)
      if @hangar!=nil and i>=0
        @shieldBoosters << @hangar.removeShieldBooster(i)
      end
    end

    # Equipa el arma de índice 'i' del hangar a la estación
    def mountWeapon(i)
      if @hangar!=nil and i>=0
        @weapons << @hangar.removeWeapon(i)
      end
    end

    # Movimiento de la estación
    def move
      if @fuelUnits>0
        @fuelUnits = [0,@fuelUnits - @fuelUnits*getSpeed].max
      end
    end
    
    # Calcula el poder de escudo de la nave
    def protection
      factor = 1.0
      @shieldBoosters.each do |shield|
        factor = factor*shield.useIt
      end
      factor*@shieldPower
    end

    # Recibe un hangar
    def recieveHangar(h)
      if @hangar==nil
        @hangar = h
      end
    end

    # Añade un escudo al hangar en caso de ser posible
    def recieveShieldBooster(s)
      if @hangar!=nil
        @hangar.addShieldBooster(s)
      else
        return false
      end
    end

    # Añade un arma al hangar en caso de ser posible
    def recieveWeapon(w)
      if @hangar!=nil
        @hangar.addWeapon(w)
      else
        return false
      end
    end
    
    # Determina si la nave aguanta un disparo
    def recieveShot(shot)
      myProtection = protection
      if myProtection>=shot
        @shieldPower = [0.0,@shieldPower - shot*@@SHIELDLOSSPERUNITSHOT].max
        return ShotResult::RESIST
      else
        @shieldPower = 0.0
        return ShotResult::DONOTRESIST
      end
    end

    # La nave recibe suministros
    def recieveSupplies(s)
      @ammoPower += s.ammoPower
      @fuelUnits = [@@MAXFUEL,@fuelUnits+s.fuelUnits].min
      @shieldPower += s.shieldPower
    end

    # Recepción de un botín
    def setLoot(loot)
      dealer = CardDealer.instance

      h = loot.nHangars
      if h>0
        hangar = dealer.nextHangar
        recieveHangar(hangar)
      end

      elements = loot.nSupplies
      for i in 0...elements 
        sup = dealer.nextSuppliesPackage
        recieveSupplies(sup)
      end

      elements = loot.nWeapons
      for i in 0...elements
        weap = dealer.nextWeapon
        recieveWeapon(weap)
      end

      elements = loot.nShields
      for i in 0...elements
        sh = dealer.nextShieldBooster
        recieveShieldBooster(sh)
      end

      medals = loot.nMedals
      @nMedals += medals

      if loot.getEfficient==true
        return Transformation::GETEFFICIENT
      elsif loot.spaceCity==true
        return Transformation::SPACECITY
      else
        return Transformation::NOTRANSFORM
      end
    end

    # Ajusta el  daño a la nave y lo guarda
    def setPendingDamage(d)
      @pendingDamage = d.adjust(@weapons,@shieldBoosters)
    end

    # Comprueba si el estado de la nave es válido
    def validState
      if @pendingDamage==nil or @pendingDamage.hasNoEffect
        return true
      else
        return false
      end
    end

    def getUIversion
      SpaceStationToUI.new(self)
    end

    def to_s
      getUIversion.to_s
    end
  end
end

File: pdoo/java/Deepspace/SpaceStation.java
/**
 * @author Manuel Gachs Ballegeer
 */
package Deepspace;

import java.util.ArrayList;

/**
 * @brief Clase estación espacial
 */
public class SpaceStation implements SpaceFighter {
    private final static int MAXFUEL = 100;
    private final static float SHIELDLOSSPERUNITSHOT = 0.1f;
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals = 0;
    private float shieldPower;
    private Damage pendingDamage;
    private Hangar hangar;
    private ArrayList<Weapon> weapons = new ArrayList<>();
    private ArrayList<ShieldBooster> shieldBoosters = new ArrayList<>();
    
    SpaceStation(String n,SuppliesPackage supplies) {
        hangar = null;
        pendingDamage = null;
        name = n;
        ammoPower = supplies.getAmmoPower();
        shieldPower = supplies.getShieldPower();
        assignFuelValue(supplies.getFuelUnits());
    }
    
    SpaceStation(SpaceStation station) {
        hangar = station.hangar;
        pendingDamage = station.pendingDamage;
        name = station.name;
        ammoPower = station.ammoPower;
        shieldPower = station.shieldPower;
        assignFuelValue(station.fuelUnits);
    }
    
    private void assignFuelValue(float f) {
        fuelUnits = (f<MAXFUEL)?f:MAXFUEL;
    }
    
    private void cleanPendingDamage() {
        if (pendingDamage!=null && pendingDamage.hasNoEffect())
            pendingDamage = null;
    }
    
    public void cleanUpMountedUnits() {
        for (Weapon w: weapons)
            if (w.getUses()==0)
                weapons.remove(w);
        for (ShieldBooster s: shieldBoosters)
            if (s.getUses()==0)
                shieldBoosters.remove(s);
    }
    
    public void discardHangar() {
        if (hangar!=null)
            hangar = null;
    }
    
    public void discardShieldBooster(int i) {
        if (i>=0 && i<shieldBoosters.size()) {
            if (pendingDamage!=null)
                pendingDamage.discardShieldBooster();
            shieldBoosters.remove(i);
        }
        
    }
    
    public void discardShieldBoosterInHangar(int i) {
        if (hangar!=null)
            hangar.removeShieldBooster(i);
    }
    
    public void discardWeapon(int i) {
        if (i>=0 && i<weapons.size()) {
            if (pendingDamage!=null) {
                Weapon w = weapons.get(i);
                pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }
            weapons.remove(i);
        }
    }
    
    public void discardWeaponInHangar(int i) {
        if (hangar!=null)
            hangar.removeWeapon(i);
    }
    
    @Override
    public float fire() {
        float factor = 1.0f;
        for (Weapon w: weapons)
            factor = factor*w.useIt();
        return factor;
    }
    
    public float getAmmoPower() { return ammoPower; }
    
    public float getFuelUnits() { return fuelUnits;  }
    
    public Hangar getHangar() { return hangar; }
    
    public String getName() { return name; }
    
    public int getNMedals() { return nMedals; }
    
    public Damage getPendingDamage() { return pendingDamage; }
    
    public ArrayList<ShieldBooster> getShieldBoosters() { return shieldBoosters; }
    
    public float getShieldPower() { return shieldPower; }
    
    public float getSpeed() { return fuelUnits/MAXFUEL; }
    
    SpaceStationToUI getUIversion() { return new SpaceStationToUI(this); }
    
    public ArrayList<Weapon> getWeapons() { return weapons; }
    
    public void mountShieldBooster(int i) {
        ShieldBooster aux = null;
        if (hangar!=null) {
            aux = hangar.removeShieldBooster(i);
            if (aux!=null)
                shieldBoosters.add(aux);
        }
    }
    
    public void mountWeapon(int i) {
        Weapon aux = null;
        if (hangar!=null) {
            aux = hangar.removeWeapon(i);
            if (aux!=null)
                weapons.add(aux);
        }
    }
    
    public void move() { fuelUnits = ((fuelUnits-getSpeed())<0)?0:(fuelUnits-getSpeed()); }
    
    @Override
    public float protection() {
        float factor = 1.0f;
        for (ShieldBooster s: shieldBoosters)
            factor = factor*s.useIt();
        return factor;
    }
    
    public void recieveHangar(Hangar h) {
        if (hangar==null)
            hangar = h;
    }
    
    public boolean recieveShieldBooster(ShieldBooster s) {
        if (hangar!=null)
            return hangar.addShieldBooster(s);
        else
            return false;
    }
    
    @Override
    public ShotResult recieveShot(float shot) {
        float my_protection = protection();
        if (my_protection>=shot) {
            shieldPower = ((shieldPower-shot*SHIELDLOSSPERUNITSHOT)<0.0f)?0.0f:shieldPower-shot*SHIELDLOSSPERUNITSHOT;
            return ShotResult.RESIST;
        } else {
            shieldPower = 0.0f;
            return ShotResult.DONOTRESIST;
        }
    }
    
    public void recieveSupplies(SuppliesPackage s) {
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        assignFuelValue(fuelUnits+s.getFuelUnits());
    }
    
    public boolean recieveWeapon(Weapon w) {
        if (hangar!=null)
            return hangar.addWeapon(w);
        else
            return false;
    }
    
    public Transformation setLoot(Loot loot) {
        CardDealer dealer = CardDealer.getInstance();
        
        int h = loot.getNHangars();
        if (h>0) {
            Hangar hangar = dealer.nextHangar();
            recieveHangar(hangar);
        }
        
        int elements = loot.getNSupplies();
        for (int i=0;i<elements;i++) {
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            recieveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
        for (int i=0;i<elements;i++) {
            Weapon weap = dealer.nextWeapon();
            recieveWeapon(weap);
        }
        
        elements = loot.getNShields();
        for (int i=0;i<elements;i++) {
            ShieldBooster sh = dealer.nextShieldBooster();
            recieveShieldBooster(sh);
        }
        
        int medals = loot.getNMedals();
        nMedals += medals;
        
        if (loot.getEfficient())
            return Transformation.GETEFFICIENT;
        else if(loot.spaceCity())
            return Transformation.SPACECITY;
        else
            return Transformation.NOTRANSFORM;
    }
    
    public void setPendingDamage(Damage d) { pendingDamage = d.adjust(weapons, shieldBoosters); }
    
    public boolean validState() { return pendingDamage==null || pendingDamage.hasNoEffect(); }
   
    @Override
    public String toString() {
        String aux = "Nombre de la nave: " + name;
        aux += "\nCombustible: " + fuelUnits;
        aux += ", Potencia de disparo: " + ammoPower;
        aux += ", Potencia de escudo: " + shieldPower;
        aux += "\nMedallas: " + nMedals;
        aux += "\n--------------------------\nHangar:";
        if (hangar!=null)
            aux += hangar.toString();
        aux += "--------------------------\nArmas:\n";
         for (Weapon w: weapons)
            aux += w.toString() + "\n";
        aux += "Escudos:\n";
        for (ShieldBooster s: shieldBoosters)
            aux += s.toString() + "\n";
        aux += "Daño: ";
            if (pendingDamage!=null)
                aux += pendingDamage.toString() + "\n";
        return aux;
    }
    
}